controller/routes_test.go

```diff
@@ -347,3 +349,234 @@ func (s *S) TestStreamRouteEvents(c *C) {
+	app2 := s.createTestApp(c, &api.App{DisplayName: "app2"})
+	app3 := s.createTestApp(c, &api.App{DisplayName: "app3"})
+	var app1Routes, app2Routes, app3Routes []*api.Route
+
+	// define some convenience functions
+	httpRoute := func(service, domain string) *api.Route {
+		return &api.Route{
+			ServiceTarget: &api.Route_ServiceTarget{
+				ServiceName:   service,
+				DrainBackends: true,
+			},
+			Config: &api.Route_Http{Http: &api.Route_HTTP{
+				Domain: domain,
+				Path:   "/",
+			}},
+		}
+	}
+	setRoutes := func(dryRun bool, expectedState []byte) (*api.SetRoutesResponse, error) {
+		req := &api.SetRoutesRequest{
+			AppRoutes: []*api.AppRoutes{
+				{App: app1.Name, Routes: app1Routes},
+				{App: app2.Name, Routes: app2Routes},
+				{App: app3.Name, Routes: app3Routes},
+			},
+			DryRun:        dryRun,
+			ExpectedState: expectedState,
+		}
+		return s.routerClient.SetRoutes(context.Background(), req)
+	}
+	dryRun := func() *api.SetRoutesResponse {
+		res, err := setRoutes(true, nil)
+		c.Assert(err, IsNil)
+		c.Assert(res.DryRun, Equals, true)
+		c.Assert(res.AppliedToState, Not(IsNil))
+		return res
+	}
+	applyRoutes := func(expectedState []byte) *api.SetRoutesResponse {
+		res, err := setRoutes(false, expectedState)
+		c.Assert(err, IsNil)
+		c.Assert(res.DryRun, Equals, false)
+		c.Assert(res.AppliedToState, Not(IsNil))
+		return res
+	}
+	assertChange := func(change *api.RouteChange, action api.RouteChange_Action, before, after *api.Route) {
+		c.Assert(change.Action, Equals, action)
+
+		if before == nil {
+			c.Assert(change.Before, IsNil)
+		} else {
+			c.Assert(change.Before, Not(IsNil))
+			c.Assert(change.Before.String(), Equals, before.String())
+		}
+
+		if after == nil {
+			c.Assert(change.After, IsNil)
+		} else {
+			c.Assert(change.After, Not(IsNil))
+			c.Assert(change.After.String(), Equals, after.String())
+		}
+	}
+	assertNoRoutes := func() {
+		var count int64
+		err := s.db.QueryRow(
+			"SELECT COUNT(*) FROM http_routes WHERE parent_ref IN ($1, $2, $3) AND deleted_at IS NULL",
+			"controller/"+app1.Name, "controller/"+app2.Name, "controller/"+app3.Name,
+		).Scan(&count)
+		c.Assert(err, IsNil)
+		if count != 0 {
+			c.Fatalf("expected no routes, got %d", count)
+		}
+	}
+	assertRoutes := func(expected ...*api.Route) {
+		rows, err := s.db.Query(
+			"SELECT service, domain FROM http_routes WHERE parent_ref IN ($1, $2, $3) AND deleted_at IS NULL ORDER BY domain",
+			"controller/"+app1.Name, "controller/"+app2.Name, "controller/"+app3.Name,
+		)
+		c.Assert(err, IsNil)
+		defer rows.Close()
+		var routes []*router.Route
+		for rows.Next() {
+			var route router.Route
+			c.Assert(rows.Scan(&route.Service, &route.Domain), IsNil)
+			routes = append(routes, &route)
+		}
+		c.Assert(rows.Err(), IsNil)
+		if len(routes) != len(expected) {
+			c.Fatalf("expected %d routes, got %d", len(expected), len(routes))
+		}
+		for i, r := range expected {
+			c.Assert(routes[i].Service, Equals, r.ServiceTarget.ServiceName)
```

Comment:
Should `DrainBackends` also be compared here?

Reply:
Handled in cebac8ed.